Upgrade redis

Owen-CH-Leung opened the pull request:
Upgrade redis to 4.5.5. This PR fixes: #29175

passed all the existing 5 tests for the redis provider 
```bash
breeze testing tests --test-type "Providers[redis]"
```
<img width="894" alt="image" src="https://github.com/apache/airflow/assets/43698890/e0a26d12-0c91-492a-a67e-aa75df20b8fa">


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Owen-CH-Leung committed:
Upgrade redis

Upgrade redis to 4.5.5

uranusjr commented on airflow/providers/redis/provider.yaml:
```diff
@@ -39,7 +39,7 @@ dependencies:
   # https://github.com/redis/redis-py/blob/master/CHANGES
   # TODO: upgrade to support redis package >=4
-  - redis~=3.2
+  - redis~=4.5.5
```
Can 3.x still be supported? Also `~=4.5.5` limits the version range to only `>=4.5.5,<4.6.0`, which is likely not the intention?

Owen-CH-Leung replied:
Thanks for your comment. How does `redis>=3.2,<4.5.5` sound to you ?

pankajkoti replied:
do we wish to have an upper bound? can we not keep just `>=3.2`
I do not have enough context on this. Do we know if there was there a reason that we were not able to support `redis >= 3.3.0` and now we're able to do so?  Do we have enough tests in our provider to test the `mixins` and unquoting URLS with `urllib.parse.unquote` as mentioned in the comment?

Owen-CH-Leung replied:
I think the existing test suites isn't enough to cover the concerns mentioned in the comment. I'll try to enrich that in this PR

potiuk replied:
Yes `>=3.2.0' is best. As long as we are reasonably sure that our tests are covering basic integraiton, not having an upper-binding is strongly preferred - the automated constraint management and upgrading to latest released version (and our canary builds failing in case of problems detected) work best in this case. 

> I think the existing test suites isn't enough to cover the concerns mentioned in the comment. I'll try to enrich that in this PR

To be honest - I am not even sure if that comment is relevant. So **maybe** just checking that it is not relevant is enough. I think it was put there out of abundance of caution. I think our usage of redis in the operators is very limited so just having an actually executed and tested example dag with our current redis would be more than enough to consider this "cool". 

Another thing we should consider together with that test is to bring redis server in our tests to later version,  https://github.com/apache/airflow/blob/main/scripts/ci/docker-compose/integration-celery.yml#LL32C1-L33C1 
Redis 5 we have does not get security fixes any more https://endoflife.date/redis - which means EOL.

I would consider bumping it to at least 6.2 version.

Our integration tests are more of "smoke" tests so we do not comprehensively test integration with multiple versions of the software we run, but theoretically we could even consider running matrix of tests for 6.0, 6.2, 7.0 of redis as well.

Owen-CH-Leung replied:
After some digging, I can confirm that the use of redis 4 wouldn't break the existing code base and should be safe to upgrade. The fact that it passes the celery integration test is a hint that it should be ok. 

For the concerns mentioned in the comment: 
- **mixins in redis commands** : This has no impact to airflow since `RedisHook` didn't directly use `redis.client`. If we are using `redis.client`, we'd need to switch to `redis.command` 
- **unquoting URLS with `urllib.parse.unquote`** : This has no impact to airflow as well since we are not using the `from_url` function 

I've just revised the doc and added a example dag to illustrate the use of the existing 3 redis Operator / sensor. I ran pytest in breeze environment & can confirm that the DAG is good to go

Owen-CH-Leung committed:
Add example dag and revise doc

Owen-CH-Leung committed:
fixing failing static check

eladkal commented on airflow/providers/redis/provider.yaml:
```diff
@@ -39,7 +39,13 @@ dependencies:
   # as well as unquoting URLS with `urllib.parse.unquote`:
   # https://github.com/redis/redis-py/blob/master/CHANGES
   # TODO: upgrade to support redis package >=4
-  - redis~=3.2
```
Please remove the comment block above if no longer relevant

Owen-CH-Leung committed:
Update provider.yaml

potiuk commented on airflow/providers/redis/provider.yaml:
```diff
@@ -35,11 +35,12 @@ versions:
+  # Changes in Redis 4 only brings into impact when we are directly importing redis.client
+  # to execute the commands, which we are not doing.
+  # The issue with unquoting URLS with `urllib.parse.unquote` is also not applicable
+  # to airflow, because we are not using the from_url function
```
```suggestion
  # Upgrading redis package >= 4 doesn't require code changes
  # Changes in Redis 4 only brings into impact when we are directly importing redis.client
  # to execute the commands, which we are not doing.
  # The issue with unquoting URLS with `urllib.parse.unquote` is also not applicable
  # to airflow, because we are not using the from_url function
```

I think what @eladkal ment (and Also what I think) - you can remove the whole comment now and simply make  it part of commit message. There is no point in leaving in the code, but the commit message will carry the trace of this being verified.

eladkal replied:
yeah sorry for not being clear.
If version is OK no need to further explain why it's OK.
We explain only things that we need to consider for future work

Owen-CH-Leung replied:
Thanks all. I removed the comments and make it as part of commit msg. I'm fixing the static check error and the build-doc error that I get from the CI tests. Will let you guys know once they are fixed.

potiuk commented on tests/system/providers/redis/example_redis_publish.py:
```diff
@@ -0,0 +1,97 @@
+#
```
This is cool! Thanks for being receptive :)

Owen-CH-Leung committed:
Remove Comment, Fix Formatting & Sphinx error

Upgrading redis package >= 4 doesn't require code changes. Changes in Redis 4 only brings into impact when we are directly importing redis.client to execute the commands, which we are not doing. The issue with unquoting URLS with urllib.parse.unquote is also not applicable to airflow, because we are not using the from_url function
